File: 33_Search In Rotated Sorted Array.java
// Binary Search (One Pass)
// Time complexity: O(logn)
// Space complexity: O(1)

class Solution {
    public int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            // case: find the target
            if (nums[mid] == target) return mid;

            // [mid, right] sorted
            if (nums[mid] < nums[right]) {
                // check if target in [mid, right]
                if (nums[mid] < target && target <= nums[right]) left = mid + 1;
                else right = mid - 1;
            }
            // [left, mid] sorted
            else {
                // check if target in [left, mid]
                if (nums[left] <= target && target < nums[mid]) right = mid - 1;
                else left = mid + 1;
            }
        }

        return -1;
    }
}

File: 33_Search In Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # [greater, smaller]
        l, r = 0, len(nums) - 1
        while l < r:
            m = (l + r) >> 1

            if nums[m] == target: return m

            # Case: in greater range
            # nums[m] > target:
            #   nums[l] <= target: [l, m-1]
            #   nums[l] > target: [m+1, r]
            # nums[m] < target: [m+1, r]
            if nums[m] > nums[r]:
                if nums[l] <= target < nums[m]: r = m - 1
                else: l = m + 1
            # Case: in smaller range
            # nums[m] > target: [l, m-1]
            # nums[m] < target: 
            #   nums[r] < target: [l, m-1]
            #   nums[r] >= target: [m+1, r]
            else:
                if nums[m] < target <= nums[r]: l = m + 1
                else: r = m - 1
        
        return l if nums[l] == target else -1